### PR TITLE
[wait-for] lambda support

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following is a very initial idea for adding lambda support for
each query.

The idea is that we have a series of scopes that a scope can add to when
we enter a lambda. This is kind of like stack registers, but not as
complicated. The reason we don't need to implement stack registers is
that we're using the go runtime to do it for us, so we can piggyback
on to that without having to do any of the work!

Additional work to change Ord to Box. The Box value is more generalized
and not only talks about Ord syntax but syntax we need to do other
things like accessor values.

The idea of Box is the same as Java, except there is no bottom type.

Previously I was aiming to do `applications.forEach`, but actually that
introduces a lot of complexity into the query evaluator, so much so it
makes it really hard to understand. Instead, nothing can have a method
and we use high order functions to do this for us. We don't get the partial
application, but I don't think we need it tbh.

The code is rather simple, lift the sub-entity into its own scope and
then use that when we call the lambda. We get a lot of things for free
and it works surprisingly well.

The accessor is identical to the infix in terms of using the order
precedence, but we have a new AST type so that the query evaluator
doesn't become a mess.

    juju-wait-for model default --query='forEach(applications, app => app.status == "active")'

As there is no order in the forEach, it might be wise to order them to
get some valid expected behavior.

## QA steps

```sh
juju-wait-for model default --query='forEach(applications, app => app.status == "active")'
```

### CODE REVIEW

File: cmd/plugins/juju-wait-for/strategy.go [hunk 1]
```diff
@@ -156,23 +177,21 @@ func (m GenericScope) GetIdentValue(name string) (query.Ord, error) {
...
-// GetIdents returns the identifers that are supported for a given scope.
-func (m GenericScope) GetIdents() []string {
-	var res []string
-
-	refType := reflect.TypeOf(m.Info).Elem()
-	for i := 0; i < refType.NumField(); i++ {
-		field := refType.Field(i)
-		v := strings.Split(field.Tag.Get("json"), ",")[0]
-		refValue := reflect.ValueOf(m.Info).Elem()
+// SetIdentValue sets a nwe ident and it's value on a given scope.
```
Comment (manadart) [inline cmd/plugins/juju-wait-for/strategy.go:183]: "new"

**Review by manadart — COMMENTED** [review]
If I create the model, run the query, then deploy ubuntu, the query doesn't return even when it becomes active.

**Review by manadart — APPROVED** [review]
Works well now.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**